File: config_io_module.f90
! config_io_module.f90
! Routines for atomic/molecular configuration I/O
MODULE config_io_module

  !------------------------------------------------------------------------------------------------!
  ! This software was written in 2016/17                                                           !
  ! by Michael P. Allen <m.p.allen@warwick.ac.uk>/<m.p.allen@bristol.ac.uk>                        !
  ! and Dominic J. Tildesley <d.tildesley7@gmail.com> ("the authors"),                             !
  ! to accompany the book "Computer Simulation of Liquids", second edition, 2017 ("the text"),     !
  ! published by Oxford University Press ("the publishers").                                       !
  !                                                                                                !
  ! LICENCE                                                                                        !
  ! Creative Commons CC0 Public Domain Dedication.                                                 !
  ! To the extent possible under law, the authors have dedicated all copyright and related         !
  ! and neighboring rights to this software to the PUBLIC domain worldwide.                        !
  ! This software is distributed without any warranty.                                             !
  ! You should have received a copy of the CC0 Public Domain Dedication along with this software.  !
  ! If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.                               !
  !                                                                                                !
  ! DISCLAIMER                                                                                     !
  ! The authors and publishers make no warranties about the software, and disclaim liability       !
  ! for all uses of the software, to the fullest extent permitted by applicable law.               !
  ! The authors and publishers do not recommend use of this software for any purpose.              !
  ! It is made freely available, solely to clarify points made in the text. When using or citing   !
  ! the software, you should not imply endorsement by the authors or publishers.                   !
  !------------------------------------------------------------------------------------------------!

  USE, INTRINSIC :: iso_fortran_env, ONLY : error_unit, iostat_end, iostat_eor

  IMPLICIT NONE
  PRIVATE

  ! Public routines
  PUBLIC :: read_cnf_atoms, write_cnf_atoms, read_cnf_mols, write_cnf_mols

CONTAINS

  SUBROUTINE read_cnf_atoms ( filename, n, box, r, v ) ! Read in atomic configuration
    IMPLICIT NONE
    CHARACTER(len=*),               INTENT(in)    :: filename ! Supplied filename
    INTEGER,                        INTENT(inout) :: n        ! Number of atoms
    REAL,                           INTENT(out)   :: box      ! Simulation box length (assumed cubic)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(out)   :: r        ! Atomic positions (3,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(out)   :: v        ! Atomic velocities (3,n)

    ! The first call of this routine is just to get n and box, used to allocate arrays
    ! The second call attempts to read in the atomic positions and optionally velocities

    INTEGER :: cnf_unit, ioerr, i

    ! Open given filename, will terminate on any errors

    OPEN ( newunit=cnf_unit, file=filename, status='old', action='read', iostat=ioerr )
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error opening ', filename, ioerr
       STOP 'Error in read_cnf_atoms'
    END IF

    ! Read number of atoms from first record
    
    READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) n
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading n from ', filename, ioerr
       IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
       IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
       STOP 'Error in read_cnf_atoms'
    END IF

    ! Read box length from second record
    
    READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) box
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading box from ', filename, ioerr
       IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
       IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
       STOP 'Error in read_cnf_atoms'
    END IF

    ! Expected format is one record per atom containing either r(:,i), v(:,i) or just r(:,i)

    IF ( PRESENT ( r ) ) THEN

       IF ( n > SIZE ( r, dim=2 ) ) THEN
          WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of r ', n, SIZE ( r, dim=2 )
          STOP 'Error in read_cnf_atoms'
       END IF

       IF ( PRESENT ( v ) ) THEN

          IF ( n > SIZE ( v, dim=2 ) ) THEN
             WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of v ', n, SIZE ( v, dim=2 )
             STOP 'Error in read_cnf_atoms'
          END IF

          ! Read positions, velocities
          DO i = 1, n
             READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) r(:,i), v(:,i)
             IF ( ioerr /= 0 ) THEN
                WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading r, v from ', filename, ioerr
                IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
                IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
                STOP 'Error in read_cnf_atoms'
             END IF
          END DO

       ELSE

          ! Read positions
          DO i = 1, n
             READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) r(:,i)
             IF ( ioerr /= 0 ) THEN
                WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading r from ', filename, ioerr
                IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
                IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
                STOP 'Error in read_cnf_atoms'
             END IF
          END DO

       END IF

    END IF

    CLOSE ( unit=cnf_unit )

  END SUBROUTINE read_cnf_atoms

  SUBROUTINE write_cnf_atoms ( filename, n, box, r, v ) ! Write out atomic configuration
    IMPLICIT NONE
    CHARACTER(len=*),               INTENT(in) :: filename ! Supplied filename
    INTEGER,                        INTENT(in) :: n        ! Number of atoms
    REAL,                           INTENT(in) :: box      ! Simulation box length (assumed cubic)
    REAL, DIMENSION(:,:),           INTENT(in) :: r        ! Atomic positions (3,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(in) :: v        ! Atomic velocities

    INTEGER :: cnf_unit, ioerr, i

    ! Open given filename, replacing it if it already exists, will terminate on any errors
    OPEN ( newunit=cnf_unit, file=filename, status='replace', iostat=ioerr )
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error opening ', filename, ioerr
       STOP 'Error in write_cnf_atoms'
    END IF

    ! Write number of atoms to first record, box length to second record
    WRITE ( unit=cnf_unit, fmt='(i15)'  ) n
    WRITE ( unit=cnf_unit, fmt='(f15.8)') box

    IF ( n > SIZE ( r, dim=2 ) ) THEN
       WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of r ', n, SIZE ( r, dim=2 )
       STOP 'Error in write_cnf_atoms'
    END IF

    IF ( PRESENT ( v ) ) THEN

       IF ( n > SIZE ( v, dim=2 ) ) THEN
          WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of v ', n, SIZE ( v, dim=2 )
          STOP 'Error in write_cnf_atoms'
       END IF

       ! Write positions, velocities
       DO i = 1, n
          WRITE ( unit=cnf_unit, fmt='(*(f15.10))' ) r(:,i), v(:,i)
       END DO

    ELSE

       ! Write positions
       DO i = 1, n
          WRITE ( unit=cnf_unit, fmt='(*(f15.10))' ) r(:,i)
       END DO

    END IF

    CLOSE ( unit=cnf_unit )

  END SUBROUTINE write_cnf_atoms

  SUBROUTINE read_cnf_mols ( filename, n, box, r, e, v, w ) ! Read in molecular configuration
    IMPLICIT NONE
    CHARACTER(len=*),               INTENT(in)    :: filename ! Supplied filename
    INTEGER,                        INTENT(inout) :: n        ! Number of molecules
    REAL,                           INTENT(out)   :: box      ! Simulation box length (assumed cubic)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(out)   :: r        ! Molecular positions (3,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(out)   :: e        ! Orientation vectors (3,n) or quaternions (4,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(out)   :: v        ! Molecular velocities (3,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(out)   :: w        ! Angular velocities/momenta (3,n)

    ! The first call of this routine is just to get n and box, used to allocate arrays
    ! The second call attempts to read in the molecular positions, orientations
    ! and optionally velocities, angular velocities/momenta

    INTEGER :: cnf_unit, ioerr, i

    ! Open given filename, will terminate on any errors

    OPEN ( newunit=cnf_unit, file=filename, status='old', action='read', iostat=ioerr )
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error opening ', filename, ioerr
       STOP 'Error opening file in read_cnf_mols'
    END IF

    ! Read number of molecules from first record
    
    READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) n
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading n from ', filename, ioerr
       IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
       IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
       STOP 'Error in read_cnf_mols'
    END IF

    ! Read box length from second record
    
    READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) box
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading box from ', filename, ioerr
       IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
       IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
       STOP 'Error in read_cnf_mols'
    END IF

    ! Expected format is one line per atom containing either r(:,i), e(:,i), v(:,i), w(:,i)  or just r(:,i), e(:,i)
    ! The first dimension of the e array can be 3 elements (vector) or 4 elements (quaternion)

    IF ( PRESENT ( r ) ) THEN

       IF ( .NOT. PRESENT ( e )    ) THEN
          WRITE ( unit=error_unit, fmt='(a,a,i15)') 'r and e arguments must be present together'
          STOP 'Error in read_cnf_mols'
       END IF
       IF ( n > SIZE ( r, dim=2 ) ) THEN
          WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of r ', n, SIZE ( r, dim=2 )
          STOP 'Error in read_cnf_mols'
       END IF
       IF ( n > SIZE ( e, dim=2 ) ) THEN
          WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of e ', n, SIZE ( e, dim=2 )
          STOP 'Error in read_cnf_mols'
       END IF

       IF ( PRESENT ( v ) ) THEN

          IF ( .NOT. PRESENT ( w )    ) THEN
             WRITE ( unit=error_unit, fmt='(a,a,i15)') 'v and w arguments must be present together'
             STOP 'Error in read_cnf_mols'
          END IF
          IF ( n > SIZE ( v, dim=2 ) ) THEN
             WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of v ', n, SIZE ( v, dim=2 )
             STOP 'Error in read_cnf_mols'
          END IF
          IF ( n > SIZE ( w, dim=2 ) ) THEN
             WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of w ', n, SIZE ( w, dim=2 )
             STOP 'Error in read_cnf_mols'
          END IF

          ! Read positions, orientation vectors or quaternions, velocities, angular velocities/momenta
          DO i = 1, n
             READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) r(:,i), e(:,i), v(:,i), w(:,i)
             IF ( ioerr /= 0 ) THEN
                WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading r, e, v, w from ', filename, ioerr
                IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
                IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
                STOP 'Error in read_cnf_mols'
             END IF
          END DO

       ELSE

          ! Read positions, orientation vectors or quaternions
          DO i = 1, n
             READ ( unit=cnf_unit, fmt=*, iostat=ioerr ) r(:,i), e(:,i)
             IF ( ioerr /= 0 ) THEN
                WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error reading r, e from ', filename, ioerr
                IF ( ioerr == iostat_eor ) WRITE ( unit=error_unit, fmt='(a)') 'End of record'
                IF ( ioerr == iostat_end ) WRITE ( unit=error_unit, fmt='(a)') 'End of file'
                STOP 'Error in read_cnf_mols'
             END IF
          END DO

       END IF

    END IF

    CLOSE ( unit=cnf_unit )

  END SUBROUTINE read_cnf_mols

  SUBROUTINE write_cnf_mols ( filename, n, box, r, e, v, w ) ! Write out molecular configuration
    IMPLICIT NONE
    CHARACTER(len=*),               INTENT(in) :: filename ! Supplied filename
    INTEGER,                        INTENT(in) :: n        ! Number of molecules
    REAL,                           INTENT(in) :: box      ! Simulation box length (assumed cubic)
    REAL, DIMENSION(:,:),           INTENT(in) :: r        ! Molecular positions (3,n)
    REAL, DIMENSION(:,:),           INTENT(in) :: e        ! Orientation vectors (3,n) or quaternions (0:3,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(in) :: v        ! Molecular velocities (3,n)
    REAL, DIMENSION(:,:), OPTIONAL, INTENT(in) :: w        ! Angular velocities/momenta (3,n)

    INTEGER :: cnf_unit, ioerr, i

    ! Open given filename, replacing it if it already exists, will terminate on any errors
    OPEN ( newunit=cnf_unit, file=filename, status='replace', iostat=ioerr )
    IF ( ioerr /= 0 ) THEN
       WRITE ( unit=error_unit, fmt='(a,a,i15)') 'Error opening ', filename, ioerr
       STOP 'Error in write_cnf_mols'
    END IF

    ! Write number of molecules to first record, box length to second record
    WRITE(cnf_unit,'(i15)'  ) n
    WRITE(cnf_unit,'(f15.8)') box

    IF ( n > SIZE ( r, dim=2 ) ) THEN
       WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of r ', n, SIZE ( r, dim=2 )
       STOP 'Error in write_cnf_mols'
    END IF
    IF ( n > SIZE ( e, dim=2 ) ) THEN
       WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of e ', n, SIZE ( e, dim=2 )
       STOP 'Error in write_cnf_mols'
    END IF

    IF ( PRESENT ( v ) ) THEN
       IF ( .NOT. PRESENT ( w )    ) THEN
          WRITE ( unit=error_unit, fmt='(a,a,i15)') 'v and w arguments must be present together'
          STOP 'Error in write_cnf_mols'
       END IF
       IF ( n > SIZE ( v, dim=2 ) ) THEN
          WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of v ', n, SIZE ( v, dim=2 )
          STOP 'Error in write_cnf_mols'
       END IF
       IF ( n > SIZE ( w, dim=2 ) ) THEN
          WRITE ( unit=error_unit, fmt='(a,2i15)') 'Error in size of w ', n, SIZE ( w, dim=2 )
          STOP 'Error in write_cnf_mols'
       END IF

       ! Write positions, orientation vectors or quaternions, velocities, angular velocities/momenta
       DO i = 1, n
          WRITE ( unit=cnf_unit, fmt='(*(f15.10))') r(:,i), e(:,i), v(:,i), w(:,i)
       END DO

    ELSE

       ! Write positions, orientation vectors or quaternions
       DO i = 1, n
          WRITE ( unit=cnf_unit, fmt='(*(f15.10))') r(:,i), e(:,i)
       END DO

    END IF

    CLOSE ( unit=cnf_unit )

  END SUBROUTINE write_cnf_mols

END MODULE config_io_module

File: python_examples/config_io_module.py
#!/usr/bin/env python3
# config_io_module.py

#------------------------------------------------------------------------------------------------#
# This software was written in 2016/17                                                           #
# by Michael P. Allen <m.p.allen@warwick.ac.uk>/<m.p.allen@bristol.ac.uk>                        #
# and Dominic J. Tildesley <d.tildesley7@gmail.com> ("the authors"),                             #
# to accompany the book "Computer Simulation of Liquids", second edition, 2017 ("the text"),     #
# published by Oxford University Press ("the publishers").                                       #
#                                                                                                #
# LICENCE                                                                                        #
# Creative Commons CC0 Public Domain Dedication.                                                 #
# To the extent possible under law, the authors have dedicated all copyright and related         #
# and neighboring rights to this software to the PUBLIC domain worldwide.                        #
# This software is distributed without any warranty.                                             #
# You should have received a copy of the CC0 Public Domain Dedication along with this software.  #
# If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.                               #
#                                                                                                #
# DISCLAIMER                                                                                     #
# The authors and publishers make no warranties about the software, and disclaim liability       #
# for all uses of the software, to the fullest extent permitted by applicable law.               #
# The authors and publishers do not recommend use of this software for any purpose.              #
# It is made freely available, solely to clarify points made in the text. When using or citing   #
# the software, you should not imply endorsement by the authors or publishers.                   #
#------------------------------------------------------------------------------------------------#

"""Routines for atomic/molecular configuration I/O."""

def read_cnf_atoms ( filename, with_v=False ):
    """Read in atomic configuration."""
    
    import numpy as np
    
    with open(filename,"r") as f:
        n=int(f.readline()) # Number of atoms
        box=float(f.readline()) # Simulation box length (assumed cubic)
    rv=np.loadtxt(filename,skiprows=2) # The rest of the file should be uniformly formatted
    rows, cols = rv.shape
    assert rows == n, "{:d}{}{:d}".format(rows,' rows not equal to ',n)
    assert cols >= 3, "{:d}{}".format(cols,' cols less than 3')
    r = rv[:,0:3].astype(np.float_) # Coordinate array
    if with_v:
        assert cols >= 6, "{:d}{}".format(cols,' cols less than 6')
        v = rv[:,3:6].astype(np.float_) # Velocity array
        return n, box, r, v
    else:
        return n, box, r

def read_cnf_mols ( filename, with_v=False, quaternions=False ):
    """Read in molecular configuration."""
    
    import numpy as np
    
    with open(filename,"r") as f:
        n=int(f.readline()) # Number of atoms
        box=float(f.readline()) # Simulation box length (assumed cubic)
    revw=np.loadtxt(filename,skiprows=2) # The rest of the file should be uniformly formatted
    rows, cols = revw.shape
    assert rows == n, "{:d}{}{:d}".format(rows,' rows not equal to ',n)
    cols_re = 7 if quaternions else 6
    assert cols >= cols_re, "{:d}{}{:d}".format(cols,' cols less than ',cols_re)
    r = revw[:,0:3].astype(np.float_) # Coordinate array
    e = revw[:,3:cols_re].astype(np.float_) # Orientation array
    if with_v:
        assert cols >= cols_re+6, "{:d}{}{:d}".format(cols,' cols less than',cols_re+6)
        v = revw[:,cols_re  :cols_re+3].astype(np.float_) # Velocity array
        w = revw[:,cols_re+3:cols_re+6].astype(np.float_) # Angular velocity/momentum array
        return n, box, r, e, v, w
    else:
        return n, box, r, e

def write_cnf_atoms ( filename, nn, box, *args ):
    """Write out atomic configuration."""

    import numpy as np

    nargs=len(args)
    assert nargs==1 or nargs==2, "{}{:d}".format('Wrong number of arguments ',nargs)

    my_header="{:15d}\n{:15.8f}".format(nn,box)

    n, d = args[0].shape
    assert n==nn, "r shape mismatch {:d}{:d}".format(n,nn)
    assert d==3,  "r shape mismatch {:d}".format(d)

    if nargs==1: # Positions only
        r=args[0] # Just r
        np.savetxt(filename,r,header=my_header,fmt='%15.10f',comments='')

    else: # Positions and velocities
        n, d = args[1].shape
        assert n==nn, "v shape mismatch {:d}{:d}".format(n,nn)
        assert d==3, "v shape mismatch {:d}".format(d)
        rv=np.concatenate((args[0],args[1]),axis=1) # Both r and v
        np.savetxt(filename,rv,header=my_header,fmt='%15.10f',comments='')

def write_cnf_mols ( filename, nn, box, *args ):
    """Write out molecular configuration."""

    import numpy as np

    nargs=len(args)
    assert nargs==2 or nargs==4, "{}{:d}".format('Wrong number of arguments ',nargs)

    my_header="{:15d}\n{:15.8f}".format(nn,box)

    n, d = args[0].shape # Positions
    assert n==nn, "r shape mismatch {:d}{:d}".format(n,nn)
    assert d==3,  "r shape mismatch {:d}".format(d)

    n, d = args[1].shape # Orientations: vectors or quaternions
    assert n==nn, "e shape mismatch {:d}{:d}".format(n,nn)
    assert d==3 or d==4, "e shape mismatch {:d}".format(d)

    if nargs==2: # Positions and orientations
        re=np.concatenate((args[0],args[1]),axis=1) # Just r and e
        np.savetxt(filename,re,header=my_header,fmt='%15.10f',comments='')

    else: # Positions, orientations, velocities and angular velocities/momenta
        n, d = args[2].shape
        assert n==nn, "v shape mismatch {:d}{:d}".format(n,nn)
        assert d==3, "v shape mismatch {:d}".format(d)
        n, d = args[3].shape
        assert n==nn, "w shape mismatch {:d}{:d}".format(n,nn)
        assert d==3, "w shape mismatch {:d}".format(d)
        revw=np.concatenate((args[0],args[1],args[2],args[3]),axis=1) # All of r, e, v and w
        np.savetxt(filename,revw,header=my_header,fmt='%15.10f',comments='')
